NATIVE-486: Add documentation for CI/CD (#255)
Integreat/integreat-react-native-app: docs/06-manual-builds.md

# Manual builds

> If you want to create `.ipa` or `.apk` builds for testing purposes you can follow this guide. If you only want to quickly push an update to the stores then take a look [at the CI/CD documentation](08-cicd.md#triggering-a-build-in-ci).

Reviewer: ```suggestion
If you want to run the app locally or create `.ipa` or `.apk` builds for testing purposes you can follow this guide. If you only want to quickly push an update to the stores then take a look [at the CI/CD documentation](08-cicd.md#triggering-a-build-in-ci).
```

> ## Manually building for iOS

Reviewer: Imo this section is too complicated if I just want to run it locally on my device/on the emulator (which should be the main usecase here). At least the section to run it in the emulator/local device should be first imo.

Reviewer: I think the corresponding commands should also be part of getting started of the readme. If I am a new contributor to the project, I don't want to know all the details of fastlane, certificates.... but rather just run the app and start developing.

Author: I tried to restructure it a bit. But it still is a little bit complicated.

I already linked it from the README.

First setup XCode and [CocoaPods](#setup-of-cocoapods).
Now you have to prepare a [few environment variables](08-cicd.md#environment-variables-and-dependencies):

```bash
export FASTLANE_USER=<secret>
export FASTLANE_PASSWORD=<secret>
export MATCH_PASSWORD=<secret>
```

Setup the certificates locally:

```bash
cd ios/
bundle exec fastlane certificates
```

> *Hint: After setting up the certificates you can also star to use XCode to build the app.*

Reviewer: Do we have an example on how to do this?
```suggestion
*Note: After setting up the certificates you can also start to use XCode to build the app.*
```

Build the app:

```bash
yarn
cd ios/

> bundle exec pod install

Reviewer: Imo this should be moved to setup since those are duplicates.

### Run the app in a simulator

First setup XCode and [CocoaPods](#setup-of-cocoapods).
If you want to quickly run the app in an emulator do the following:

```bash
yarn
cd ios/
bundle exec pod install
cd ..
yarn start
yarn ios
```

### Setup of CocoaPods

CocoaPods is used for dependency management of the native libraries.
First you have to make sure that Ruby is working correctly. We advise to use RVM as mentioned in the [setup of Fastlane](08-cicd.md#setup-of-fastlane). After running this setup you should also be able to do `bundle exec pod --version`.

*Note: `bundle exec pod install` uses the versions from the `Podfile.lock`. `bundle exec pod update` updates the `Podfile.lock`.*

## Manually building for Android

If you do not have Fastlane installed [skip to the next section](#quick-builds-using-a-test-signing-key).
Firstly you have to prepare a [few environment variables](08-cicd.md#environment-variables-and-dependencies):

```bash
export CREDENTIALS_GIT_REPOSITORY_URL=<secret>
export CREDENTIALS_DIRECTORY_PATH=/tmp/credentials
export CREDENTIALS_KEYSTORE_PASSWORD=<secret>
export CREDENTIALS_KEYSTORE_PATH=/tmp/credentials/<secret>.enc

> export KEYSTORE_PATH=/tmp/keystore.jks

Reviewer: I think it would be better not having to define environment variables but rather passing them as arguments to fastlane commands.

Author: I created https://issues.integreat-app.de/browse/NATIVE-548 instead

```

Setup your JKS now:

```bash
cd android/
bundle exec fastlane keystore
```

Finally, provide some more information about unlocking the JKS:

```bash
export KEYSTORE_KEY_ALIAS=<secret>
export KEYSTORE_PASSWORD=<secret>
export KEYSTORE_KEY_PASSWORD=<secret>
```

> The last step is to build to app:

Reviewer: ```suggestion
The last step is to build the app:
```

```bash
yarn
cd android/
bundle exec fastlane build
```

Install and run the app using:

```bash
adb install app/build/outputs/apk/release/app-release.apk
adb shell am force-stop tuerantuer.app.integreat
adb shell am start -n tuerantuer.app.integreat/.MainActivity
```

### Quick builds using a test signing key

If you don't want to deal with the production signing key you can use the test singing JKS.

Optionally, you can explicitly set the environment variables:

```bash
export ORG_GRADLE_PROJECT_KEYSTORE_PATH=test.keystore
export ORG_GRADLE_PROJECT_KEYSTORE_PASSWORD=123456
export ORG_GRADLE_PROJECT_KEYSTORE_KEY_ALIAS=test
export ORG_GRADLE_PROJECT_KEYSTORE_KEY_PASSWORD=123456
```

Optionally, you can also set the version name and code:

```bash
export ORG_GRADLE_PROJECT_VERSION_CODE=1
export ORG_GRADLE_PROJECT_VERSION_NAME=0.1
```

> Then you can create a quick rest build in release mode and run it on your emulator:

Reviewer: I'd also move this before any optional setup. As mentioned above, imo this is too complicated.

> Then you can create a quick rest build in release mode and run it on your emulator:

Reviewer: Also running the device locally on my device is completely missing.

## Manually building for iOS

### Prerequisites
* First setup XCode and [CocoaPods](#setup-of-cocoapods).
* In order to set up certificates you also need [to set up Fastlane](08-cicd.md#setup-of-fastlane).
* Read about [Apple Certificates](10-apple-certifcates.md)
* Setup CocoaPods

  CocoaPods is used for dependency management of the native libraries.
  First you have to make sure that Ruby is working correctly. We advise to use RVM as mentioned in the [setup of Fastlane](08-cicd.md#setup-of-fastlane). After running this setup you should also be able to do `bundle exec pod --version`.
    
  *Note: `bundle exec pod install` uses the versions from the `Podfile.lock`. `bundle exec pod update` updates the `Podfile.lock`.*

### Setup Certificates

*Note: This step is required in order to build the project for iOS*

First you have to prepare a [few environment variables](08-cicd.md#environment-variables-and-dependencies):

```bash
export FASTLANE_USER=<secret>
export FASTLANE_PASSWORD=<secret>
export MATCH_PASSWORD=<secret>
```

Finally, install the certificates locally:

```bash
cd ios/
bundle exec fastlane certificates

> ```

Author: I favor not to use bullet lists here as I never did so far

> *Note: After setting up the certificates you can also start to use XCode to build the app.*

Author: This is a good tip imo

### Run the app in a Simulator

```bash
yarn
cd ios/
bundle exec pod install
cd ..
yarn start
yarn ios
```

### Run the app on a Device or Simulator via XCode

```bash
yarn
cd ios/
bundle exec pod install
cd ..
yarn start
```

Now start XCode and open `ios/Integreat.xcworkspace`. Then you can build and start the app on your device or simulator.

### Building the App with a Distribution Certificate
Build the app:

```bash
yarn
cd ios/
bundle exec pod install
bundle exec fastlane build
```

Fastlane should report where the build artifacts are. These can be uploaded to App Store Connect or distributed via another way.

## Manually building for Android

If you do not have Fastlane installed [you can skip to the next section](#quick-builds-using-a-test-signing-key), which does not require FL.

> 

Author: I restructured the section

### Prerequisites
* Setup of the Android SDK
* In order to set up production keystores you also need [to set up Fastlane](08-cicd.md#setup-of-fastlane).

### Building the App for Play Store

Firstly you have to prepare a [few environment variables](08-cicd.md#environment-variables-and-dependencies):

```bash
export CREDENTIALS_GIT_REPOSITORY_URL=<secret>
export CREDENTIALS_DIRECTORY_PATH=/tmp/credentials
export CREDENTIALS_KEYSTORE_PASSWORD=<secret>
export CREDENTIALS_KEYSTORE_PATH=/tmp/credentials/<secret>.enc
export KEYSTORE_PATH=/tmp/keystore.jks
```

Setup your JKS now:

```bash
cd android/
bundle exec fastlane keystore
```

> Finally, provide some more information about unlocking the JKS:

Reviewer: I'd add these to prepare environment variables above.

Author: I'd keep these separate as it makes clear that the first step decrypts the keystore and is not required for the `fastlane keystore` command

Author: I restructured the section

```bash
export KEYSTORE_KEY_ALIAS=<secret>
export KEYSTORE_PASSWORD=<secret>

> export KEYSTORE_KEY_PASSWORD=<secret>

Reviewer: I'd move this to a separate sub like in the ios part. I'd also apply the suggestions made for the ios certificate setup part.

Author: I restructured the section

```

The last step is to build the app:

```bash
yarn
cd android/
bundle exec fastlane build
```

Install and run the app using:

```bash
adb install app/build/outputs/apk/release/app-release.apk
adb shell am force-stop tuerantuer.app.integreat

> adb shell am start -n tuerantuer.app.integreat/.MainActivity

Reviewer: Why is this necessary?

Author: These are valuable tricks to quickly restart the app

```

### Quick builds using a test signing key

If you don't want to deal with the production signing key you can use the test singing JKS.

Optionally, you can explicitly set the environment variables:

```bash
export ORG_GRADLE_PROJECT_KEYSTORE_PATH=test.keystore
export ORG_GRADLE_PROJECT_KEYSTORE_PASSWORD=123456
export ORG_GRADLE_PROJECT_KEYSTORE_KEY_ALIAS=test
export ORG_GRADLE_PROJECT_KEYSTORE_KEY_PASSWORD=123456
```

Optionally, you can also set the version name and code:

```bash
export ORG_GRADLE_PROJECT_VERSION_CODE=1
export ORG_GRADLE_PROJECT_VERSION_NAME=0.1
```

Then you can create a quick test build in release mode and run it on your emulator:

```bash
yarn
yarn android:release
```

> ### Run the app in an Emulator
> 
> If you want to quickly run the app in an emulator do the following:

Reviewer: ```suggestion
### Run the app in an Emulator or device

If you want to quickly run the app in an emulator or connected device do the following:
```
